### ISSUE TITLE
TraceSamplingRules fails to load

### ISSUE THREAD

**Issue by Jtalk** [DataDog/dd-trace-java#6494]
TraceSamplingRules is failing on startup with ExceptionInInitializerError due to what seems to be a misconfigured Moshi adapter. 

They have a negative [test](https://github.com/square/moshi/blob/4d3ca9a1746f7425d33bc9b2000656b0bdfa392b/moshi/src/test/java/com/squareup/moshi/AdapterMethodsTest.java#L524) that checks against [exactly what RuleAdapter seems to be doing](https://github.com/DataDog/dd-trace-java/blob/b9ad8e52dee14c42320bb1e993757a60d9c2720a/dd-trace-core/src/main/java/datadog/trace/common/sampling/TraceSamplingRules.java#L157).

It currently fails for me in v1.27.0.

```
[dd.trace 2024-01-15 16:26:14:412 +0000] [main] ERROR datadog.trace.bootstrap.Agent - Throwable thrown while installing the Datadog Tracer
java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at datadog.trace.bootstrap.Agent.installDatadogTracer(Agent.java:582)
        at datadog.trace.bootstrap.Agent.access$400(Agent.java:68)
        at datadog.trace.bootstrap.Agent$InstallDatadogTracerCallback.execute(Agent.java:468)
        at datadog.trace.bootstrap.Agent.start(Agent.java:300)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at datadog.trace.bootstrap.AgentBootstrap.agentmain(AgentBootstrap.java:71)
        at datadog.trace.bootstrap.AgentBootstrap.premain(AgentBootstrap.java:55)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:491)
        at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndCallPremain(InstrumentationImpl.java:503)
Caused by: java.lang.ExceptionInInitializerError
        at datadog.trace.agent.common.sampling.Sampler$Builder.forConfig(Sampler.java:37)
        at datadog.trace.agent.core.CoreTracer$CoreTracerBuilder.config(CoreTracer.java:423)
        at datadog.trace.agent.core.CoreTracer$CoreTracerBuilder.<init>(CoreTracer.java:412)
        at datadog.trace.agent.core.CoreTracer.builder(CoreTracer.java:123)
        at datadog.trace.agent.tooling.TracerInstaller.installGlobalTracer(TracerInstaller.java:21)
        ... 20 more
Caused by: java.lang.IllegalArgumentException: No @ToJson adapter for class datadog.trace.agent.common.sampling.TraceSamplingRules$Rule (with no annotations)
for class datadog.trace.agent.common.sampling.TraceSamplingRules$Rule
for java.util.List<datadog.trace.agent.common.sampling.TraceSamplingRules$Rule>
        at com.squareup.moshi.Moshi$LookupChain.exceptionWithLookupStack(Moshi.java:388)
        at com.squareup.moshi.Moshi.adapter(Moshi.java:157)
        at com.squareup.moshi.Moshi.adapter(Moshi.java:105)
        at com.squareup.moshi.Moshi.adapter(Moshi.java:74)
        at com.squareup.moshi.CollectionJsonAdapter.newArrayListAdapter(CollectionJsonAdapter.java:54)
        at com.squareup.moshi.CollectionJsonAdapter$1.create(CollectionJsonAdapter.java:38)
        at com.squareup.moshi.Moshi.adapter(Moshi.java:145)
        at com.squareup.moshi.Moshi.adapter(Moshi.java:105)
        at com.squareup.moshi.Moshi.adapter(Moshi.java:74)
        at datadog.trace.agent.common.sampling.TraceSamplingRules.<clinit>(TraceSamplingRules.java:24)
        ... 25 more
Caused by: java.lang.IllegalArgumentException: No @ToJson adapter for class datadog.trace.agent.common.sampling.TraceSamplingRules$Rule (with no annotations)
        at com.squareup.moshi.AdapterMethodsFactory.create(AdapterMethodsFactory.java:60)
        at com.squareup.moshi.Moshi.adapter(Moshi.java:145)
        ... 33 more
Caused by: java.lang.IllegalArgumentException: cannot construct instances of datadog.trace.agent.common.sampling.TraceSamplingRules$Rule
        at com.squareup.moshi.ClassFactory.get(ClassFactory.java:145)
        at com.squareup.moshi.ClassJsonAdapter$1.create(ClassJsonAdapter.java:103)
        at com.squareup.moshi.Moshi.nextAdapter(Moshi.java:176)
        at com.squareup.moshi.AdapterMethodsFactory.create(AdapterMethodsFactory.java:53)
        ... 34 more

```

**Comment by nayeem-kamal** [comment]
@Jtalk Thank you for submitting your issue. Could you please provide me with the rules you are setting so that I can reproduce the issue.

**Comment by Jtalk** [comment]
Hey @nayeem-kamal, I'm not setting any rules. It's failing in `Moshi.adapter` in the static initialiser. I only have the standard set (DD_SERVICE, DD_ENV, DD_VERSION, DD_AGENT_HOST), as well as `DD_TRACE_ENABLED=true` and `DD_PROFILING_ENABLED=true`, but, judging by how it's crashing in `Agent.installDatadogTracer`, I'd guess that simply enabling the tracer would suffice.

**Comment by ygree** [comment]
Released in [v1.29.0](https://github.com/DataDog/dd-trace-java/releases/tag/v1.29.0 )
